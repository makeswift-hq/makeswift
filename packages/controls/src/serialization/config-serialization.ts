import { match, P } from 'ts-pattern'
import {
  isFunction,
  isSerializedFunction,
  serializeFunction,
  deserializeFunction,
  type AnyFunction,
} from './function-serialization'

import { mapValues } from '../utils/functional'

export function serializeConfig(config: unknown): [unknown, Transferable[]] {
  const transferables: Transferable[] = []

  const serializeFunc = (func: AnyFunction) => {
    const r = serializeFunction(func)
    transferables.push(r)
    return r
  }

  const serialize = (value: unknown): unknown =>
    match(value)
      .with(P.when(isFunction), serializeFunc)
      .with(P.array(), (arr) => arr.map(serialize))
      .with({}, (obj) => mapValues(obj, serialize))
      .otherwise(() => value)

  return [serialize(config), transferables]
}

export function deserializeConfig(config: unknown): unknown {
  const deserialize = (value: unknown): unknown =>
    match(value)
      .with(P.when(isSerializedFunction), deserializeFunction)
      .with(P.array(), (arr) => arr.map(deserialize))
      .with({}, (obj) => mapValues(obj, deserialize))
      .otherwise(() => value)

  return deserialize(config)
}
