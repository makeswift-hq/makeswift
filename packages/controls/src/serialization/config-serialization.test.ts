import { serializeConfig, deserializeConfig } from './config-serialization'

expect.addSnapshotSerializer({
  serialize: () => '[MessagePort]',
  test: (value) => value instanceof MessagePort,
})

describe('config serialization', () => {
  describe.each([
    {},
    { label: undefined },
    { label: 'visible', defaultValue: true },
    { label: 'visible', options: [{ label: 'one', value: 1 }] },
  ])('%o', (config) => {
    test('should serialize to itself', () => {
      expect(serializeConfig(config)).toEqual([config, []])
    })

    test('should deserialize to itself', () => {
      const [serialized, _] = serializeConfig(config)
      expect(deserializeConfig(serialized)).toEqual(config)
    })
  })

  describe.each([
    { getItemLabel: (item: any) => item.label },
    { arrayOfFunctions: [(item: any) => item.label] },
  ])('%o', (config) => {
    test('should handle function serialization', async () => {
      const [serialized, transferables] = serializeConfig(config)
      expect(transferables).toHaveLength(1)
      expect(transferables[0]).toBeInstanceOf(MessagePort)
      transferables.forEach((port: any) => port.close())

      expect(serialized).toMatchSnapshot()
    })

    test('should deserialize serialized functions', () => {
      const [serialized, transferables] = serializeConfig(config)
      transferables.forEach((port: any) => port.close())
      expect(deserializeConfig(serialized)).toMatchSnapshot()
    })
  })
})
