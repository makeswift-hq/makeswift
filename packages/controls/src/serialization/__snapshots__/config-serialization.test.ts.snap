// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`config serialization {
  arrayOfFunctions: [ [Function (anonymous)] { [length]: 1, [name]: '' }, [length]: 1 ]
} should deserialize serialized functions 1`] = `
{
  "arrayOfFunctions": [
    [Function],
  ],
}
`;

exports[`config serialization {
  arrayOfFunctions: [ [Function (anonymous)] { [length]: 1, [name]: '' }, [length]: 1 ]
} should handle function serialization 1`] = `
{
  "arrayOfFunctions": [
    [MessagePort],
  ],
}
`;

exports[`config serialization {
  getItemLabel: [Function: getItemLabel] { [length]: 1, [name]: 'getItemLabel' }
} should deserialize serialized functions 1`] = `
{
  "getItemLabel": [Function],
}
`;

exports[`config serialization {
  getItemLabel: [Function: getItemLabel] { [length]: 1, [name]: 'getItemLabel' }
} should handle function serialization 1`] = `
{
  "getItemLabel": [MessagePort],
}
`;
